packages/file_selector/file_selector_android/android/src/main/java/dev/flutter/packages/file_selector_android/FileUtils.java

```diff
@@ -0,0 +1,198 @@
+import android.net.Uri;
+import android.os.Environment;
+import android.provider.DocumentsContract;
+import android.provider.MediaStore;
+import android.webkit.MimeTypeMap;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.UUID;
+
+public class FileUtils {
+
+  /** URI authority that represents access to external storage providers. */
+  public static final String EXTERNAL_DOCUMENT_AUTHORITY = "com.android.externalstorage.documents";
+
+  /**
+   * Retrieves path of directory represented by the specified {@code Uri}.
+   *
+   * <p>Intended to handle any cases needed to return paths from URIS retrieved from open
+   * documents/directories by starting one of {@code Intent.ACTION_OPEN_FILE}, {@code
+   * Intent.ACTION_OPEN_FILES}, or {@code Intent.ACTION_OPEN_DOCUMENT_TREE}.
+   *
+   * <p>Will return the path for on-device directories, but does not handle external storage
+   * volumes.
+   */
+  @NonNull
+  public static String getPathFromUri(@NonNull Context context, @NonNull Uri uri) {
+    String uriAuthority = uri.getAuthority();
+
+    if (uriAuthority.equals(EXTERNAL_DOCUMENT_AUTHORITY)) {
+      String uriDocumentId = DocumentsContract.getDocumentId(uri);
+      String documentStorageVolume = uriDocumentId.split(":")[0];
+
+      // Non-primary storage volumes come from SD cards, USB drives, etc. and are
+      // not handled here.
+      if (!documentStorageVolume.equals("primary")) {
+        throw new UnsupportedOperationException(
+            "Retrieving the path of a document from storage volume "
+                + documentStorageVolume
+                + " is unsupported by this plugin.");
+      }
+      String innermostDirectoryName = uriDocumentId.split(":")[1];
+      String externalStorageDirectory = Environment.getExternalStorageDirectory().getPath();
+
+      return externalStorageDirectory + "/" + innermostDirectoryName;
+    } else {
+      throw new UnsupportedOperationException(
+          "Retrieving the path from URIs with authority "
+              + uriAuthority.toString()
+              + " is unsupported by this plugin.");
+    }
+  }
+
+  /**
+   * Copies the file from the given content URI to a temporary directory, retaining the original
+   * file name if possible.
+   *
+   * <p>Each file is placed in its own directory to avoid conflicts according to the following
+   * scheme: {cacheDir}/{randomUuid}/{fileName}
+   *
+   * <p>File extension is changed to match MIME type of the file, if known. Otherwise, the extension
+   * is left unchanged.
+   *
+   * <p>If the original file name is unknown, a predefined "file_selector" filename is used and the
+   * file extension is deduced from the mime type.
+   *
+   * <p>Will return null if closing the output stream when done copying file contents does not for
```

Comment:
I'm not sure I follow the first case here "when done copying file contents does not for sure complete", can you clarify?